File: ocaml/bin/day02.ml
let read_input filename = 
  let in_channel = open_in filename in
  let lines = In_channel.input_lines in_channel in
  close_in in_channel;

  let parse_line (line: string): int list = 
    let numbers = String.split_on_char ' ' line in
    List.map (fun s -> int_of_string s) numbers
  in

  List.map (fun line -> parse_line line) lines


let is_report_safe (report: int list): bool = 

  let rec valid_pair pair increasing =
    match ((fst pair - snd pair), increasing) with 
    | (diff, _) when abs diff > 3 || abs diff == 0 -> false
    | (diff, incr) -> if incr then diff < 0 else diff > 0
  in

  let rec iterate l increasing =
    match (l, increasing) with 
    | ([], _) -> true
    | (h1::h2::t, None) -> 
      let incr = h1 < h2 in
      let pair_valid = if incr then valid_pair (h1, h2) incr else valid_pair (h1, h2) false in
      if pair_valid then iterate (h2::t) (Some incr) else false
    | (h1::h2::t, Some incr) -> 
      let pair_valid = if incr then valid_pair (h1, h2) incr else valid_pair (h1, h2) false in
      if pair_valid then iterate (h2::t) (Some incr) else false
  in

  iterate report None


let silver = 
  let input = read_input "../inputs/02.txt" in

  let res = List.fold_left (fun sum v -> sum + (if is_report_safe v then 1 else 0)) 0 input in

  print_int res


let () =
  silver

File: python/day02.py
def read_input(filename: str) -> list[list[int]]:
    with open(filename, "r") as file:
        return [[int(word) for word in line.strip().split()] for line in file]


def is_report_safe(report: list[int]) -> bool:
    incr: bool | None = None
    for i in range(len(report) - 1):
        diff = report[i] - report[i + 1]
        incr = incr if incr is not None else diff < 0
        if abs(diff) > 3 or abs(diff) == 0:
            return False
        if (incr and diff > 0) or (not incr and diff < 0):
            return False

    return True


def is_report_safe_gold(report: list[int]) -> bool:
    if is_report_safe(report):
        return True

    for i in range(len(report)):
        cp = list.copy(report)
        cp.pop(i)
        if is_report_safe(cp):
            return True

    return False


def silver() -> int:
    input = read_input("../inputs/02.txt")
    return sum([int(is_report_safe(report)) for report in input])


def gold() -> int:
    input = read_input("../inputs/02.txt")
    return sum([int(is_report_safe_gold(report)) for report in input])


if __name__ == "__main__":
    print(f"Silver: {silver()}")
    print(f"Gold: {gold()}")
